packages/headless/src/app/renew-access-token-middleware.ts

```diff
@@ -0,0 +1,56 @@
+import {Middleware} from '@reduxjs/toolkit';
+import {debounce} from 'ts-debounce';
+import {updateBasicConfiguration} from '../features/configuration/configuration-actions';
+import {ExpiredTokenError} from '../utils/errors';
+
+export function createRenewAccessTokenMiddleware(
+  renewToken: () => Promise<string>
+): Middleware {
+  let accessTokenRenewalsAttempts = 0;
+  const resetRenewalTriesAfterDelay = debounce(
+    () => (accessTokenRenewalsAttempts = 0),
+    500
+  );
+
+  return (store) => (next) => async (action) => {
+    const isThunk = typeof action === 'function';
+
+    if (!isThunk) {
+      return next(action);
+    }
+
+    const payload = await next(action);
+
+    if (!hasExpiredTokenError(payload)) {
+      return payload;
+    }
+
+    accessTokenRenewalsAttempts++;
+    resetRenewalTriesAfterDelay();
+
+    if (accessTokenRenewalsAttempts > 5) {
+      return payload;
+    }
+
+    const accessToken = await attempt(renewToken);
+
+    if (!accessToken) {
+      return payload;
+    }
+
+    store.dispatch(updateBasicConfiguration({accessToken}));
+    return next(action);
+  };
```

Comment:
I has trouble unit testing the middleware, both directly and when configuring it on a real or mock store.

One option could be to extract and export just the renewal portion, but I felt this would be not very informative. I opted to use an integration test instead, since the middleware relies on the platform client throwing a specific error.